docs/user-guide/pods.md

```diff
@@ -110,6 +110,18 @@ Pod is exposed as a primitive in order to facilitate:
 
 The current best practice for pets is to create a replication controller with `replicas` equal to `1` and a corresponding service. If you find this cumbersome, please comment on [issue #260](http://issue.k8s.io/260).
 
+
+## Termination of Pods
+
+Because pods represent running processes on nodes in the cluster, it is important to allow those processes to gracefully terminate when they are no longer needed (vs being violently killed with a KILL signal and having no chance to clean up). Users should be able to request deletion and know when processes terminate, but also be able to ensure that deletes eventually complete. Kubernetes represents a distributed system, so it's not always possible to observe graceful shutdown of a process, and in some cases processes may not cleanly shut down due to kernel or system issues. When a user requests deletion of a pod the system records the intended grace period before the pod is allowed to be forcefully killed, and a TERM signal is sent to the main process in each container. Once the grace period has expired the KILL signal is sent to those processes and the pod is then deleted from the API server. If processes from that pod are still running, the system will work to terminate them.
+
+By default, all deletes are graceful within 30 seconds. Commands support the `--grace-period=<seconds>` option which allows a user to override the default and specify their own value. The value `0` indicates that delete should be immediate, and removes the pod in the API immediately so a new pod can be created with the same name. On the node pods that are set to terminate immediately will still be given a small grace period before being force killed.
+
+Deleted pods will be shown as "Terminating" when listed in the client commands.
```

Comment:
Sure

On Aug 15, 2015, at 3:01 AM, David Oppenheimer <notifications@github.com>
wrote:

In docs/user-guide/pods.md
<https://github.com/kubernetes/kubernetes/pull/12747#discussion_r37134183>:

> @@ -110,6 +110,18 @@ Pod is exposed as a primitive in order to facilitate:
>
>  The current best practice for pets is to create a replication controller with `replicas` equal to `1` and a corresponding service. If you find this cumbersome, please comment on [issue #260](http://issue.k8s.io/260).
>
> +
> +## Termination of Pods
> +
> +Because pods represent running processes on nodes in the cluster, it is important to allow those processes to gracefully terminate when they are no longer needed (vs being violently killed with a KILL signal and having no chance to clean up). Users should be able to request deletion and know when processes terminate, but also be able to ensure that deletes eventually complete. Kubernetes represents a distributed system, so it's not always possible to observe graceful shutdown of a process, and in some cases processes may not cleanly shut down due to kernel or system issues. When a user requests deletion of a pod the system records the intended grace period before the pod is allowed to be forcefully killed, and a TERM signal is sent to the main process in each container. Once the grace period has expired the KILL signal is sent to those processes and the pod is then deleted from the API server. If processes from that pod are still running, the system will work to ter
 minate t
 hem.
> +
> +By default, all deletes are graceful within 30 seconds. Commands support the `--grace-period=<seconds>` option which allows a user to override the default and specify their own value. The value `0` indicates that delete should be immediate, and removes the pod in the API immediately so a new pod can be created with the same name. On the node pods that are set to terminate immediately will still be given a small grace period before being force killed.
> +
> +Deleted pods will be shown as "Terminating" when listed in the client commands.

I found this sentence and the next paragraph confusing. Maybe you can write
down a sequence of events, like
1. User sends command to delete Pod, with default grace period
2. Processes in the Pod are immediately given SIGTERM
3. (simultaneous with 2) Pod immediately shows up as "Terminating" when
listed in client commands (I'm not sure if this is right -- you said
*deleted* pods show up as Terminating, but I am assuming you mean *deleting*
because it seems like it wouldn't make sense to call an already-deleted pod
"Terminating")
4. (simultaneous with 2), Pod is removed from endpoints list for service,
and are no longer considered part of the set of running pods for
replication controllers. Pods that shutdown slowly can continue to serve
traffic as load balancers (like the service proxy) remove them from their
rotations.
5. When the grace period expires, any processes still running in the Pod
are killed with SIGKILL.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12747/files#r37134183>.